src/dispatch.c

```diff
@@ -82,3 +83,452 @@ natsStatus natsSub_enqueueUserMessage(natsSubscription *sub, natsMsg *msg)
+    bool isRealMessage = (msg->subject[0] != '\0');
+    if (isRealMessage)
+    {
+        sub->ownDispatcher.queue.msgs--;
+        sub->ownDispatcher.queue.bytes -= natsMsg_dataAndHdrLen(msg);
+    }
+
+    // Fetch-specific handling of synthetic and header-only messages
+    if (fetch != NULL)
+        fetchStatus = js_checkFetchedMsg(sub, msg, true, userMsg);
+
+    // Is it another kind of synthetic message?
+    *userMsg = *userMsg && (msg->subject[0] != '\0');
+
+    // Check the limits.
+    if (*userMsg)
+    {
+        if (sub->max > 0)
+        {
+            *overLimit = (sub->delivered == sub->max);
+            *lastMessageInSub = (sub->delivered == (sub->max - 1));
+        }
+
+        if (fetch)
+        {
+            bool overMaxBytes = ((fetch->lifetime.MaxBytes > 0) && ((fetch->deliveredBytes) > fetch->lifetime.MaxBytes));
+            bool overMaxFetch = ((fetch->deliveredMsgs >= fetch->lifetime.Batch) || overMaxBytes);
+
+            *lastMessageInFetch = (fetch->deliveredMsgs == (fetch->lifetime.Batch - 1) || overMaxBytes);
+
+            // See if we want to override fetch status based on our own data.
+            if (fetchStatus == NATS_OK)
+            {
+                if (*lastMessageInFetch || overMaxFetch)
+                {
+                    fetchStatus = NATS_MAX_DELIVERED_MSGS;
+                }
+                if (overMaxBytes)
+                {
+                    fetchStatus = NATS_LIMIT_REACHED;
+                }
+            }
+            *overLimit = (*overLimit || overMaxFetch || overMaxBytes);
+            *lastMessageInSub = (*lastMessageInSub || *lastMessageInFetch);
+        }
+
+        if (!*overLimit)
+        {
+            sub->delivered++;
+            if (fetch)
+            {
+                fetch->deliveredMsgs++;
+                fetch->deliveredBytes += natsMsg_dataAndHdrLen(msg);
+            }
+        }
+
+        *fcReply = (jsi == NULL ? NULL : jsSub_checkForFlowControlResponse(sub));
+    }
+
+    return fetchStatus;
+}
+
+// Thread main function for a thread pool of dispatchers.
+void nats_dispatchThreadPool(void *arg)
+{
+    natsDispatcher *d = (natsDispatcher *)arg;
+
+    nats_lockDispatcher(d);
+
+    while (true)
+    {
+        natsMsg *msg = NULL;
+        char *fcReply = NULL;
+        bool timerNeedReset = false;
+        bool userMsg = true;
+        bool timeout = false;
+        bool overLimit = false;
+        bool lastMessageInSub = false;
+        bool lastMessageInFetch = false;
+        natsStatus fetchStatus = NATS_OK;
+
+        while (((msg = d->queue.head) == NULL) && !d->shutdown)
+            natsCondition_Wait(d->cond, d->mu);
+
+        // Break out only when list is empty
+        if ((msg == NULL) && d->shutdown)
+        {
+            break;
+        }
+
+        _removeFromDispatchQueue(d, msg);
+
+        // Get subscription reference from message and capture values we need
+        // while under lock.
```

Comment:
Not the sub's lock as far as I can tell.

Reply:
Right, but all these should be protected by the dispatcher lock. This goes back to https://github.com/nats-io/nats.c/blob/v3.8.2/src/nats.c#L1698. Do you find the comment ambiguous, or do you see an issue with the code?

(Separate thread for closed/drained flags noted)